src/app.rs

```diff
@@ -217,24 +248,46 @@ impl App {
             None => return Err(ServerError::IdentityCommitmentNotFound),
         };
 
-        let proof = merkle_tree.proof(identity_index);
-        proof.map_or(Err(ServerError::IndexOutOfBounds), |proof| {
-            Ok(InclusionProofResponse {
-                root: merkle_tree.root(),
-                proof,
-            })
-        })
+        let proof = merkle_tree
+            .proof(identity_index)
+            .ok_or(ServerError::IndexOutOfBounds)?;
+        let root = merkle_tree.root();
+
+        // Locally check the proof
+        // TODO: Check the leaf index / path
+        if !merkle_tree.verify(*identity_commitment, &proof) {
+            error!(
+                ?identity_commitment,
+                ?identity_index,
+                ?root,
+                "Proof does not verify locally."
+            );
+            panic!("Proof does not verify locally.");
+        }
+
+        // Verify the root on chain
+        if let Err(error) = self.contracts.assert_valid_root(root).await {
+            error!(
+                computed_root = ?root,
+                ?error,
+                "Root mismatch between tree and contract."
+            );
+            panic!("Root mismatch between tree and contract.");
```

Comment:
Asking for inclusion proof for an outdated root (see above comment), panics the whole sequencer.